### PR TITLE
Feature/new self codes

### PR DESCRIPTION
**Pull request by ajshedivy**
### Issues:

 - #137 

#### TODO

- [x] better check for os version and ptf levels, see https://github.com/codefori/vscode-db2i/blob/c1e7da1ed85fa242844889bb74b0e5867607461b/src/config.ts#L21 (Liam - met with Scott, he said current logic should work ok)
- [x] Self Code default enablement (Liam)
- [x] Use decoration provider for hit count in SELF Code view instead of in the node description (we already do this for Dove written by Dave!)
- [x] Configuration option to have auto-refresh enabled/disabled automatically
- [x] Make use of `viewsWelcome` in `package.json` for when there are no SELF code errors.
- [x] Fix SQL statement to clear SELF logs: 
image
- [x] Button on the SELF error to copy the SQL statement to the clipboard
- [x] Show reason code when it's not null
- [x] Update SELF code tests
- [x] Change titles of SELF quick pick and view to full name
    * Quick pick: 'Set logging level for SELF'
    * View: 'SQL Error Log Facility (SELF)'
- [x] Add button in SELF view to open the docs: https://www.ibm.com/docs/en/i/7.5?topic=tools-sql-error-logging-facility-self

#### Things to test:

- [x] Ensure self codes are stored into saved settings
- [x] Ensure when starting a job from configuration, SELF codes from the settings are applied automatically
- [x] Works under *PGMR auth on the user profile (e.g. not *ALLOBJ) - scott to test

### CODE REVIEW

File: src/connection/sqlJob.ts [hunk 1]
```diff
@@ -274,10 +274,9 @@ export class SQLJob {
     return rpy;
   }
 
-  async setSelfCodes(codes: string[]) {
-    const signedCodes: String[] = codes.map(code => [code, `-${code}`]).flat();
+  async setSelfCodes(codes: string) {
     try {
-      const query: string = `SET SYSIBMADM.SELFCODES = SYSIBMADM.VALIDATE_SELF('${signedCodes.join(', ')}')`
+      const query: string = `SET SYSIBMADM.SELFCODES = '${codes}'`
       await this.query<any>(query).run();
```
Comment (worksofliam) [inline src/connection/sqlJob.ts:280]: Consider using the second parameter to `query`, which accepts binding parameters in place of `?` in the statement here.

File: src/views/jobManager/jobManagerView.ts [hunk 2]
```diff
@@ -185,20 +190,29 @@ export class JobManagerView implements TreeDataProvider<any> {
...
-          const content = `SELECT * FROM QSYS2.SQL_ERROR_LOG WHERE JOB_NAME = '${selected.job.id}'`;
-
-          vscode.commands.executeCommand(`vscode-db2i.runEditorStatement`, {
-            content,
-            qualifier: `statement`,
-            open: false,
-          });
-        }
-      }),
+      // vscode.commands.registerCommand(`vscode-db2i.jobManager.getSelfErrors`, async (node?: SQLJobItem) => {
```
Comment (worksofliam) [inline src/views/jobManager/jobManagerView.ts:193]: Remove this large comment block of code. It will be in history if we need it

**Review by worksofliam — CHANGES_REQUESTED** [review]
Just a couple of small bits. Looks good so far. I will update your todo list with a couple more items that I want to tackle here.

File: src/views/jobManager/selfCodes/selfCodesResultsView.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+  getTreeItem(element: any): TreeItem | Thenable<TreeItem> {
+    return element;
+  }
+  async getChildren(element?: any): Promise<any[]> {
+    if (element) {
+      return [];
+    } else {
+      const selfCodes = await this.getSelfCodes();
+      return selfCodes.map((error) => {
```
Comment (worksofliam) [inline src/views/jobManager/selfCodes/selfCodesResultsView.ts:84]: There is a chance that `selfCodes` can be undefined here, as `getSelfCodes` can return `undefined` if no job is running. Considering returning an empty array so then no items will show up, and then use `viewsWelcome` in the `package.json` file to have default text.

Comment (worksofliam) [reply src/views/jobManager/selfCodes/selfCodesResultsView.ts:84]: I added the if statement to make sure it is defined!

File: src/views/jobManager/selfCodes/selfCodesResultsView.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    public readonly details: string,
+    public readonly hoverMessage: vscode.MarkdownString,
+    public readonly collapsibleState: vscode.TreeItemCollapsibleState,
+    error: SelfCodeNode
+  ) {
+    super(errorMessage, collapsibleState);
+    this.tooltip = hoverMessage; // Hover text
+    this.tooltip.isTrusted = true;
+    this.description = details; // Additional details shown in the tree view
+    this.resourceUri = vscode.Uri.parse(`selfCodeTreeView:${encodeURIComponent(error.MATCHES.toString())}`);
```
Comment (worksofliam) [inline src/views/jobManager/selfCodes/selfCodesResultsView.ts:145]: Let's use `Uri.from` here instead of `parse`

File: src/extension.ts [hunk 5]
```diff
@@ -34,6 +36,10 @@ export function activate(context: vscode.ExtensionContext): Db2i {
 
   loadBase();
 
+  const exampleBrowser = new ExampleBrowser(context);
+  const selfCodesView = new selfCodesResultsView(context);
+  const selfCodeFileDecorator = new SelfTreeDecorationProvider();
```
Comment (worksofliam) [inline src/extension.ts:41]: Add this to the `context.subscriptions`

File: src/connection/sqlJob.ts (hunk 1 above)
Comment (worksofliam) [reply src/connection/sqlJob.ts:281]: We spoke with Jesse. There is a bug in the SC that stops this from working.

**Review by worksofliam — APPROVED** [review]
We have worked with Scott and he is happy. I have done a final code review, fixed some minor things, and it's looking sharp. Let's merge.

**Comment by forstie** [comment]
Thanks for all the revisions.
Finally, SELF is coming center stage.

**Pull request merged**